src/note.rs

```diff
@@ -0,0 +1,394 @@
+
+        // TODO: improve the error handling
+        cipher
+            .write(&self.encrypted_data[..])
+            .expect("Cannot write the PoseidonCipher");
+
+        let data = cipher.decrypt(&shared_secret, &nonce);
+
+        // TODO: need to figure out what we want to do in case the decryption
+        // fails
+        return if data.is_err() {
+            (0u64, JubJubScalar::zero())
+        } else {
+            let data = data.unwrap();
+            let bytes = data[0].to_bytes();
+            let value = u64::from_le_bytes(chunk_of!(8; bytes[..8]));
+            let blinding_factor =
+                JubJubScalar::from_bytes(&data[1].to_bytes()).unwrap();
+            (value, blinding_factor)
+        };
+    }
+
+    /// Create a unique nullifier for the note
+    pub fn gen_nullifier(&self, sk: &SecretSpendKey) -> BlsScalar {
+        let sk_r = sk.sk_r(&self.stealth_address);
+        let sk_r = BlsScalar::from(sk_r);
+        let pos = BlsScalar::from(self.pos());
+
+        sponge_hash(&[sk_r, pos])
+    }
+
+    /// Return a hash represented by `H(value_commitment, pos, H([R]),
+    /// H([pskr]))`
+    pub fn hash(&self) -> BlsScalar {
+        let pk_r = JubJubAffine::from(self.stealth_address.pk_r());
+
+        sponge_hash(&[
+            self.value_commitment().get_x(),
+            self.value_commitment().get_y(),
+            BlsScalar::from(self.pos()),
+            pk_r.get_x(),
+            pk_r.get_y(),
+        ])
+    }
+
+    /// Return the type of the note
+    pub fn note(&self) -> NoteType {
+        self.note_type
+    }
+
+    /// Return the position of the note on the tree.
+    pub fn pos(&self) -> u64 {
+        self.pos
+    }
+
+    /// Set the position of the note on the tree.
+    /// This, naturally, won't reflect immediatelly on the data storage
+    pub fn set_pos(&mut self, pos: u64) {
+        self.pos = pos;
+    }
+
+    /// Nonce used for the encrypt / decrypt of data for this note
+    pub fn nonce(&self) -> &JubJubScalar {
+        &self.nonce
+    }
+
+    /// Return the value commitment `H(value, blinding_factor)`
+    pub fn value_commitment(&self) -> &JubJubExtended {
+        &self.value_commitment
+    }
+
+    /// Attempt to decrypt the note value provided a [`ViewKey`]. Always
+    /// succeeds for transparent notes, and will return random values for
+    /// obfuscated notes provided the wrong view key.
+    pub fn value(&self, vk: Option<&ViewKey>) -> u64 {
+        match self.note_type {
+            NoteType::Transparent => {
+                let bytes = chunk_of!(8; &self.encrypted_data);
+                u64::from_le_bytes(bytes)
+            }
+            NoteType::Obfuscated if vk.is_some() => {
+                let (value, _) = self.decrypt_data(vk.unwrap());
+                value
+            }
+            _ => {
+                let mut rng = rand::thread_rng();
+                let value: u64 = rng.gen();
+                value
+            }
```

Comment:
Shouldn't we fail here?

Reply:
I added in the error refactoring mentioned previously also a failing here, since now we're returning a `Result`.